File: jaxtari/jaxtari/cpu/addressing.py
"""Effective-address resolution for the 6502's addressing modes.

Each `resolve_*` function takes the `CPUState` (for X / Y / PC) and a "world"
(either a `jaxtari.bus.Bus` for proper 6507 emulation, or a flat
`jnp.ndarray` for unit-test scratch memory), and returns
`(effective_addr, page_crossed, new_memory)` as plain Python ints / bools
plus a possibly-updated world. **The new_memory threading is P4d** —
RIOT reads can clear latches, so the caller has to carry the new
world forward (`memory = new_memory` after each resolver call).

`page_crossed` is only meaningful for absolute,X / absolute,Y / (indirect),Y
modes — read instructions add +1 cycle on a crossing, store instructions do not.

Memory reads are routed through `jaxtari.bus.peek` (which now also
returns `(value, new_world)`) so the same code works for both world
types and a single resolver call correctly threads any RIOT side-
effects produced during address resolution.

For implied / accumulator modes the address is meaningless and the dispatcher
in `cpu.m6502` short-circuits without calling these.

PC passed in is the address of the *opcode* itself (operand is at PC+1).
"""

from __future__ import annotations

from typing import Callable, Tuple

from jaxtari.bus.system import peek as _peek
from jaxtari.cpu.tables import (
    ADDR_ABSOLUTE,
    ADDR_ABSOLUTE_X,
    ADDR_ABSOLUTE_Y,
    ADDR_IMMEDIATE,
    ADDR_IMPLIED,
    ADDR_INDIRECT,
    ADDR_INDIRECT_X,
    ADDR_INDIRECT_Y,
    ADDR_RELATIVE,
    ADDR_ZERO,
    ADDR_ZERO_X,
    ADDR_ZERO_Y,
)
from jaxtari.types import CPUState


def _peek16(memory, addr: int):
    """Little-endian 16-bit read across two consecutive bytes (no page-wrap
    quirk). Returns `(word, new_memory)` so any side-effect bus updates
    from either byte read carry forward."""
    lo, memory = _peek(memory, addr)
    hi, memory = _peek(memory, addr + 1)
    return lo | (hi << 8), memory


def resolve_immediate(state: CPUState, memory):
    return (int(state.PC) + 1) & 0xFFFF, False, memory


def resolve_zero(state: CPUState, memory):
    operand, memory = _peek(memory, int(state.PC) + 1)
    return operand, False, memory


def resolve_zero_x(state: CPUState, memory):
    operand, memory = _peek(memory, int(state.PC) + 1)
    return (operand + int(state.X)) & 0xFF, False, memory


def resolve_zero_y(state: CPUState, memory):
    operand, memory = _peek(memory, int(state.PC) + 1)
    return (operand + int(state.Y)) & 0xFF, False, memory


def resolve_absolute(state: CPUState, memory):
    addr, memory = _peek16(memory, int(state.PC) + 1)
    return addr, False, memory


def resolve_absolute_x(state: CPUState, memory):
    """`abs,X`: effective = base + X. Task #50: on a page cross the
    real 6502 does a "wrong-page" dummy peek at (base_hi, eff_lo)
    before the corrected fetch — visible on the data bus, so the
    floating-bus latch updates."""
    base, memory = _peek16(memory, int(state.PC) + 1)
    eff = (base + int(state.X)) & 0xFFFF
    page_crossed = (base & 0xFF00) != (eff & 0xFF00)
    if page_crossed:
        wrong_addr = (base & 0xFF00) | (eff & 0xFF)
        _, memory = _peek(memory, wrong_addr)              # wrong-page dummy
    return eff, page_crossed, memory


def resolve_absolute_y(state: CPUState, memory):
    """`abs,Y`: as `abs,X` — page-cross wrong-page dummy peek."""
    base, memory = _peek16(memory, int(state.PC) + 1)
    eff = (base + int(state.Y)) & 0xFFFF
    page_crossed = (base & 0xFF00) != (eff & 0xFF00)
    if page_crossed:
        wrong_addr = (base & 0xFF00) | (eff & 0xFF)
        _, memory = _peek(memory, wrong_addr)              # wrong-page dummy
    return eff, page_crossed, memory


def resolve_indirect_x(state: CPUState, memory):
    """`(zp,X)`: pointer = (operand + X) & 0xFF, address = mem[pointer..+1] (zp-wrapped).

    Task #50: real NMOS 6502 inserts a one-cycle "internal" peek of
    `zp` (the operand byte itself, BEFORE adding X) at cycle 3,
    discarding the result. This dummy peek is visible on the data
    bus, so it updates the floating-bus latch the TIA's INPT/
    collision reads OR into D5-D0. Mirrors xitari's M6502 (zp,X)
    handling.
    """
    operand, memory = _peek(memory, int(state.PC) + 1)
    _, memory       = _peek(memory, operand)               # internal-cycle peek
    zp = (operand + int(state.X)) & 0xFF
    lo, memory = _peek(memory, zp)
    hi, memory = _peek(memory, (zp + 1) & 0xFF)
    return lo | (hi << 8), False, memory


def resolve_indirect_y(state: CPUState, memory):
    """`(zp),Y`: address = mem[zp..+1] (zp-wrapped) + Y.

    Task #50: on a page cross the real 6502 does a "wrong-page"
    dummy peek at (base_hi, eff_lo) before the corrected fetch —
    visible on the data bus, so the floating-bus latch updates.
    """
    zp, memory = _peek(memory, int(state.PC) + 1)
    lo, memory = _peek(memory, zp)
    hi, memory = _peek(memory, (zp + 1) & 0xFF)
    base = lo | (hi << 8)
    eff = (base + int(state.Y)) & 0xFFFF
    page_crossed = (base & 0xFF00) != (eff & 0xFF00)
    if page_crossed:
        wrong_addr = (base & 0xFF00) | (eff & 0xFF)
        _, memory = _peek(memory, wrong_addr)              # wrong-page dummy
    return eff, page_crossed, memory


def resolve_relative(state: CPUState, memory):
    """Branch target = PC + 2 + signed(operand). `page_crossed` reports if it crosses."""
    offset, memory = _peek(memory, int(state.PC) + 1)
    if offset >= 0x80:
        offset -= 0x100
    base = (int(state.PC) + 2) & 0xFFFF
    eff = (base + offset) & 0xFFFF
    return eff, (base & 0xFF00) != (eff & 0xFF00), memory


def resolve_indirect(state: CPUState, memory):
    """JMP indirect, faithfully replicating the 6502 page-wrap bug at $xxFF."""
    ptr, memory = _peek16(memory, int(state.PC) + 1)
    lo, memory = _peek(memory, ptr)
    hi, memory = _peek(memory, (ptr & 0xFF00) | ((ptr + 1) & 0xFF))
    return lo | (hi << 8), False, memory


Resolver = Callable[[CPUState, object], Tuple[int, bool, object]]

RESOLVERS: dict[int, Resolver] = {
    ADDR_IMMEDIATE:  resolve_immediate,
    ADDR_ZERO:       resolve_zero,
    ADDR_ZERO_X:     resolve_zero_x,
    ADDR_ZERO_Y:     resolve_zero_y,
    ADDR_ABSOLUTE:   resolve_absolute,
    ADDR_ABSOLUTE_X: resolve_absolute_x,
    ADDR_ABSOLUTE_Y: resolve_absolute_y,
    ADDR_INDIRECT:   resolve_indirect,
    ADDR_INDIRECT_X: resolve_indirect_x,
    ADDR_INDIRECT_Y: resolve_indirect_y,
    ADDR_RELATIVE:   resolve_relative,
}


INSTRUCTION_LENGTH: dict[int, int] = {
    ADDR_IMPLIED:    1,
    ADDR_IMMEDIATE:  2,
    ADDR_ZERO:       2,
    ADDR_ZERO_X:     2,
    ADDR_ZERO_Y:     2,
    ADDR_ABSOLUTE:   3,
    ADDR_ABSOLUTE_X: 3,
    ADDR_ABSOLUTE_Y: 3,
    ADDR_INDIRECT:   3,
    ADDR_INDIRECT_X: 2,
    ADDR_INDIRECT_Y: 2,
    ADDR_RELATIVE:   2,
}

File: jutari/src/cpu/Addressing.jl
"""
    Addressing

Effective-address resolution for the 6502's addressing modes. Mirrors
`jaxtari/jaxtari/cpu/addressing.py`.

Each `resolve_*` returns `(effective_addr::UInt16, page_crossed::Bool)`.
`page_crossed` only matters for absolute,X / absolute,Y / (indirect),Y modes;
loads add +1 cycle on a crossing, stores do not.
"""
module Addressing

using ..CPUTables: ADDR_IMMEDIATE, ADDR_ZERO, ADDR_ZERO_X, ADDR_ZERO_Y,
                   ADDR_ABSOLUTE, ADDR_ABSOLUTE_X, ADDR_ABSOLUTE_Y,
                   ADDR_INDIRECT, ADDR_INDIRECT_X, ADDR_INDIRECT_Y,
                   ADDR_RELATIVE, ADDR_IMPLIED
# Three dots: Types lives at JuTari.Types, but we are at JuTari.CPU.Addressing,
# so we need to climb two levels (past CPU) to reach it.
using ...Types: CPUState
# Multiple-dispatch `peek` so resolvers accept either a `BusState` (proper
# 6507 bus) or a flat `Vector{UInt8}` (P1-style scratch memory).
using ...Bus: peek as _peek
using ...Bus: pending_tick! as _tick

export resolve, instruction_length

@inline function _peek16(memory, addr::Integer)
    lo = UInt16(_peek(memory, addr))
    hi = UInt16(_peek(memory, addr + 1))
    return (hi << 8) | lo
end

@inline function _operand_addr(state::CPUState)
    return UInt16((Int(state.PC) + 1) & 0xFFFF)
end

function resolve_immediate(state::CPUState, memory)
    return _operand_addr(state), false
end

function resolve_zero(state::CPUState, memory)
    return UInt16(_peek(memory, _operand_addr(state))), false
end

function resolve_zero_x(state::CPUState, memory)
    op = _peek(memory, _operand_addr(state))
    _tick(memory)                                       # cycle-3 dummy read
    return UInt16((Int(op) + Int(state.X)) & 0xFF), false
end

function resolve_zero_y(state::CPUState, memory)
    op = _peek(memory, _operand_addr(state))
    _tick(memory)                                       # cycle-3 dummy read
    return UInt16((Int(op) + Int(state.Y)) & 0xFF), false
end

function resolve_absolute(state::CPUState, memory)
    return _peek16(memory, _operand_addr(state)), false
end

function resolve_absolute_x(state::CPUState, memory)
    # Task #50: real NMOS 6502 abs,X / abs,Y do a "wrong-page" dummy
    # peek at (base_hi, eff_lo) on page cross — visible on the data
    # bus, so the floating-bus latch updates. Mirrors jaxtari.
    base = _peek16(memory, _operand_addr(state))
    eff  = UInt16((Int(base) + Int(state.X)) & 0xFFFF)
    page = (Int(base) & 0xFF00) != (Int(eff) & 0xFF00)
    if page
        wrong = UInt16((Int(base) & 0xFF00) | (Int(eff) & 0xFF))
        _peek(memory, wrong)                                # wrong-page dummy
    end
    return eff, page
end

function resolve_absolute_y(state::CPUState, memory)
    base = _peek16(memory, _operand_addr(state))
    eff  = UInt16((Int(base) + Int(state.Y)) & 0xFFFF)
    page = (Int(base) & 0xFF00) != (Int(eff) & 0xFF00)
    if page
        wrong = UInt16((Int(base) & 0xFF00) | (Int(eff) & 0xFF))
        _peek(memory, wrong)                                # wrong-page dummy
    end
    return eff, page
end

function resolve_indirect_x(state::CPUState, memory)
    # Task #50: real NMOS 6502 (zp,X) inserts an internal-cycle peek
    # of `zp` (the operand byte before adding X) and discards the
    # result. Visible on the data bus → updates the floating-bus
    # latch. Mirrors jaxtari `resolve_indirect_x`.
    operand = _peek(memory, _operand_addr(state))
    _peek(memory, operand)                              # internal-cycle peek
    zp = UInt8((Int(operand) + Int(state.X)) & 0xFF)
    lo = UInt16(_peek(memory, zp))
    hi = UInt16(_peek(memory, UInt8((Int(zp) + 1) & 0xFF)))
    return (hi << 8) | lo, false
end

function resolve_indirect_y(state::CPUState, memory)
    # Task #50: (zp),Y page-cross wrong-page dummy peek (mirrors
    # jaxtari `resolve_indirect_y`).
    zp   = _peek(memory, _operand_addr(state))
    lo   = UInt16(_peek(memory, zp))
    hi   = UInt16(_peek(memory, UInt8((Int(zp) + 1) & 0xFF)))
    base = (hi << 8) | lo
    eff  = UInt16((Int(base) + Int(state.Y)) & 0xFFFF)
    page = (Int(base) & 0xFF00) != (Int(eff) & 0xFF00)
    if page
        wrong = UInt16((Int(base) & 0xFF00) | (Int(eff) & 0xFF))
        _peek(memory, wrong)                                # wrong-page dummy
    end
    return eff, page
end

function resolve_relative(state::CPUState, memory)
    offset = Int(_peek(memory, _operand_addr(state)))
    offset >= 0x80 && (offset -= 0x100)
    base = UInt16((Int(state.PC) + 2) & 0xFFFF)
    eff  = UInt16((Int(base) + offset) & 0xFFFF)
    return eff, (Int(base) & 0xFF00) != (Int(eff) & 0xFF00)
end

function resolve_indirect(state::CPUState, memory)
    """JMP indirect with the 6502 page-wrap bug at \$xxFF."""
    ptr = _peek16(memory, _operand_addr(state))
    lo  = UInt16(_peek(memory, ptr))
    hi  = UInt16(_peek(memory, UInt16((Int(ptr) & 0xFF00) | ((Int(ptr) + 1) & 0xFF))))
    return (hi << 8) | lo, false
end

"""
    resolve(mode, state, memory) -> (UInt16, Bool)

Dispatch to the right `resolve_*` function for the given addressing-mode code.
"""
function resolve(mode::UInt8, state::CPUState, memory)
    if mode == ADDR_IMMEDIATE;    return resolve_immediate(state, memory)
    elseif mode == ADDR_ZERO;     return resolve_zero(state, memory)
    elseif mode == ADDR_ZERO_X;   return resolve_zero_x(state, memory)
    elseif mode == ADDR_ZERO_Y;   return resolve_zero_y(state, memory)
    elseif mode == ADDR_ABSOLUTE; return resolve_absolute(state, memory)
    elseif mode == ADDR_ABSOLUTE_X; return resolve_absolute_x(state, memory)
    elseif mode == ADDR_ABSOLUTE_Y; return resolve_absolute_y(state, memory)
    elseif mode == ADDR_INDIRECT;   return resolve_indirect(state, memory)
    elseif mode == ADDR_INDIRECT_X; return resolve_indirect_x(state, memory)
    elseif mode == ADDR_INDIRECT_Y; return resolve_indirect_y(state, memory)
    elseif mode == ADDR_RELATIVE;   return resolve_relative(state, memory)
    else
        error("Addressing.resolve: implied/unknown mode $(mode)")
    end
end

"""
    instruction_length(mode) -> Int

Total instruction byte length for a given addressing mode.
"""
function instruction_length(mode::UInt8)
    if mode == ADDR_IMPLIED;      return 1
    elseif mode == ADDR_IMMEDIATE; return 2
    elseif mode == ADDR_ZERO || mode == ADDR_ZERO_X || mode == ADDR_ZERO_Y ||
           mode == ADDR_INDIRECT_X || mode == ADDR_INDIRECT_Y || mode == ADDR_RELATIVE
        return 2
    elseif mode == ADDR_ABSOLUTE || mode == ADDR_ABSOLUTE_X ||
           mode == ADDR_ABSOLUTE_Y || mode == ADDR_INDIRECT
        return 3
    else
        error("Addressing.instruction_length: unknown mode $(mode)")
    end
end

end # module
